dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -663,6 +663,322 @@ template class NaryOpNode<functional::min<double>>;
 template class NaryOpNode<std::multiplies<double>>;
 template class NaryOpNode<std::plus<double>>;
 
+// PartialReduceNode *****************************************************************
+std::vector<ssize_t> partial_reduce_shape(const std::span<const ssize_t> input_shape,
+                                          const ssize_t axis) {
+    std::vector<ssize_t> shape;
+    shape.assign(input_shape.begin(), input_shape.end());
+    shape.erase(shape.begin() + axis);
+    return shape;
+}
+
+std::vector<ssize_t> as_contiguous_strides(const std::span<const ssize_t> shape) {
+    ssize_t ndim = static_cast<ssize_t>(shape.size());
+
+    assert(ndim >= 0);
+    std::vector<ssize_t> strides(ndim);
+    // otherwise strides are a function of the shape
+    strides[ndim - 1] = sizeof(double);
+    for (auto i = ndim - 2; i >= 0; --i) {
+        strides[i] = strides[i + 1] * shape[i + 1];
+    }
+    return strides;
+}
+
+/// TODO: support multiple axes
+template <class BinaryOp>
+PartialReduceNode<BinaryOp>::PartialReduceNode(ArrayNode* node_ptr, std::span<const ssize_t> axes,
+                                               double init)
+        : ArrayOutputMixin(partial_reduce_shape(node_ptr->shape(), axes[0])),
+          init(init),
+          array_ptr_(node_ptr),
+          axes_(make_axes(axes)),
+          parent_strides_c_(as_contiguous_strides(array_ptr_->shape())) {
+    if (array_ptr_->dynamic()) {
+        throw std::invalid_argument("cannot do a partial reduction on a dynamic array");
+    } else if (array_ptr_->size() < 1) {
+        throw std::invalid_argument("cannot do a partial reduction on an empty array");
+    }
+
+    // Validate axes
+    /// TODO: support negative axes
```

Comment:
We should add an
```c++
axes.size() == 1
```
check here for now. Probably throwing an error rather than as assertion.